src/host-services.js

```diff
@@ -0,0 +1,150 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  getServicePromiseForDoc,
+  registerServiceBuilderForDoc,
+} from './service';
+
+const ServiceNames = {
+  VISIBILITY: 'host-visibility',
+  FULLSCREEN: 'host-fullscreen',
+  EXIT: 'host-exit',
+};
+
+export class HostServices {
+
+  /**
+   * @param {!Element|!./service/ampdoc-impl.AmpDoc} elementOrAmpDoc
+   * @return {!Promise<!Visibility>}
+   */
+  static visibilityForDoc(elementOrAmpDoc) {
+    return /** @type {!Promise<!Visibility>} */ (
+      getServicePromiseForDoc(elementOrAmpDoc, ServiceNames.VISIBILITY));
+  }
+
+  /**
+   * @param {!Element|!./service/ampdoc-impl.AmpDoc} elementOrAmpDoc
+   * @param {!Visibility} impl
+   */
+  static installVisibilityServiceForDoc(elementOrAmpDoc, impl) {
+    registerServiceBuilderForDoc(elementOrAmpDoc,
+        ServiceNames.VISIBILITY, impl, /* opt_instantiate */ true);
+  }
+
+  /**
+   * @param {!Element|!./service/ampdoc-impl.AmpDoc} elementOrAmpDoc
+   * @return {!Promise<!Visibility>}
+   */
+  static fullscreenForDoc(elementOrAmpDoc) {
+    return /** @type {!Promise<!Fullscreen>} */ (
+      getServicePromiseForDoc(elementOrAmpDoc, ServiceNames.FULLSCREEN));
+  }
+
+  /**
+   * @param {!Element|!./service/ampdoc-impl.AmpDoc} elementOrAmpDoc
+   * @param {!Fullscreen} impl
+   */
+  static installFullscreenServiceForDoc(elementOrAmpDoc, impl) {
+    registerServiceBuilderForDoc(elementOrAmpDoc,
+        ServiceNames.FULLSCREEN, impl, /* opt_instantiate */ true);
+  }
+
+  /**
+   * @param {!Element|!./service/ampdoc-impl.AmpDoc} elementOrAmpDoc
+   * @return {!Promise<!Exit>}
+   */
+  static exitForDoc(elementOrAmpDoc) {
+    return /** @type {!Promise<!Fullscreen>} */ (
+      getServicePromiseForDoc(elementOrAmpDoc, ServiceNames.EXIT));
+  }
+
+  /**
+   * @param {!Element|!./service/ampdoc-impl.AmpDoc} elementOrAmpDoc
+   * @param {!Exit} impl
+   */
+  static installExitServiceForDoc(elementOrAmpDoc, impl) {
+    registerServiceBuilderForDoc(elementOrAmpDoc,
+        ServiceNames.EXIT, impl, /* opt_instantiate */ true);
+  }
+}
+
+/**
+ * Visibility defines interface provided by host for visibility detection.
+ *
+ * @interface
+ */
+export class Visibility {
+
+  /**
+   * Register a callback for scroll events.
+   * @param {function(!VisibilityDataDef)} unusedCallback
+   */
+  onVisibilityChange(unusedCallback) {}
+}
+
+/**
+ * The structure that combines position and size for an element. The exact
+ * interpretation of position and size depends on the use case.
+ *
+ * @typedef {{
+ *   visibleRect: (?./layout-rect.LayoutRectDef),
```

Comment:
When pulling from MRAID's `exposureChange` event we'll have `visibleRatio` but not `visibleRect`.

Reply:
this is needed for amp-analytics to calculate visibility of elements inside the creatives. if MRAID does not support, we will have to leave it blank and report unmeasurable

Reply:
Sorry, I'm wrong!  MRAID 3's `exposureChange` callback is:

```
function(
    exposedPercentage,
    visibleRectangle,
    occlusionRectangles)
```

The idea is that you know how much of the ad is visible (`exposedPercentage`), a minimal box containing the visible area (`visibleRectangle`) and then if parts of `visibleRectangle` are not actually visible there are other rectangles to subtract from that (`occlusionRectangles`).

I need to test this, though, to make sure it works as expected on Android and iOS.

Reply:
We also need to think about the coordinate system we should use.

I suggest we take the upper-left corner of ad frame as (0,0) .